lib/active_merchant/billing/gateways/flex_charge.rb

```diff
@@ -0,0 +1,327 @@
+        }
+      end
+
+      def add_credit_card(post, credit_card, options)
+        first_name, last_name = split_names(credit_card.name)
+        post[:payment_method] = {
+          sense_key: options[:order_id],
+          credit_card: {
+            first_name: first_name,
+            last_name: last_name,
+            number: credit_card.number,
+            verification_value: credit_card.verification_value,
+            month: credit_card.month,
+            year: credit_card.year
+          }
+        }
+      end
+
+      def add_post_data(post)
+        post[:mid] = @options[:mid]
+        post[:environment] = @options[:tokenization_key] || @options[:access_token]
+      end
+
+      def evaluate(action, money, credit_card, options)
+        handle_evaluation(action, money, credit_card, options) unless options[:tokenize]
+
+        MultiResponse.run do |r|
+          r.process { tokenize(credit_card, options) }
+          extra_options = { payment_token: r.params.dig(:transaction, :payment_method) }
+          r.process { handle_evaluation(action, money, credit_card, options.merge(extra_options)) }
+        end
+      end
+
+      def handle_evaluation(action, money, credit_card, options)
+        post = {}
+        address = options[:billing_address] || options[:address]
+        add_merchant_data(post, options)
+        add_base_data(post, options)
+        add_invoice(post, money, options)
+        add_mit_data(post, money, credit_card, options)
+        add_payment(post, credit_card, address, options)
+        add_address(post, credit_card, address)
+        add_customer_data(post, options)
+        add_transaction_type(post, action)
+
+        commit(:purchase, post)
+      end
+
+      def tokenize(credit_card, options)
+        post = {}
+        add_credit_card(post, credit_card, options)
+
+        commit(:tokenize, post, headers)
+      end
+
+      def address_names(address_name, payment_method)
+        names = split_names(address_name)
+        return names if names.any?(&:present?)
+
+        [
+          payment_method&.first_name,
+          payment_method&.last_name
+        ]
+      end
+
+      def phone_from(options)
+        options[:phone] || options.dig(:billing_address, :phone) || options.dig(:billing_address, :phone_number)
+      end
+
+      def fetch_access_token
+        params = { AppKey: @options[:app_key], AppSecret: @options[:app_secret] }
+        url = build_request_url(:authenticate)
+        response = parse(ssl_post(url, params.to_json, headers))
+        @options[:access_token] = response[:accessToken]
+        @options[:expires] = response[:expires]
+      rescue ResponseError => e
+        raise OAuthResponseError.new(e)
+      end
+
+      def ensure_access_token_credential(try_again = true)
+        return if @options[:access_token] && @options[:expires] && ((@options[:expires] || 0) > DateTime.now.strftime('%Q').to_i)
```

Comment:
💬 Comment  / Style:

Maybe the double-check is not needed, you can achieve the same results with a default value:

```ruby
return if @options[:access_token] && @options.fetch(:expires, 0) > DateTime.now.strftime('%Q').to_i
```

Reply:
Done!